File: SharpMUSH.MarkupString/ColumnModule.fs
module SharpMUSH.MarkupString.ColumnModule

open System

type Justification =
    | Left
    | Center
    | Right
    | Full
    | Paragraph

[<Flags>]
type ColumnOptions =
    | Default = 0
    | Repeat = 1
    | MergeToLeft = 2
    | MergeToRight = 4
    | NoFill = 8
    | Truncate = 16
    | TruncateV2 = 32
    | NoColSep = 64

type ColumnSpec =
    { Width: int
      Justification: Justification
      Options: ColumnOptions
      // TODO: Turn string into Markup
      Ansi: string }



/// <summary>
/// Functions for parsing and handling column specifications for text alignment.
/// </summary>
module ColumnSpec =
    type private WidthPatternRegex = FSharp.Text.RegexProvider.Regex< @"^([<>=_\-])?(\d+)([\.`'$xX#]*)(?:\((.+)\))?$" >

    /// <summary>
    /// Parses a column specification string into a ColumnSpec record.
    /// </summary>
    /// <param name="spec">The column specification string.</param>
    let parse (spec: string) : ColumnSpec =
        let matchResult = WidthPatternRegex().Match(spec)

        if not matchResult.Success then
            raise (ArgumentException $"Invalid column specification: %s{spec}")

        let justification =
            if not matchResult.Groups[1].Success then
                Justification.Left
            else
                match matchResult.Groups[1].Value with
                | "<" -> Justification.Left
                | "=" -> Justification.Paragraph
                | ">" -> Justification.Right
                | "_" -> Justification.Full
                | "-" -> Justification.Center
                | _ -> Justification.Left

        let options =
            if not matchResult.Groups[3].Success then
                ColumnOptions.Default
            else
                let optionStr = matchResult.Groups[3].Value

                optionStr
                |> Seq.fold
                    (fun result c ->
                        let flag =
                            match c with
                            | '.' -> ColumnOptions.Repeat
                            | '`' -> ColumnOptions.MergeToLeft
                            | '\'' -> ColumnOptions.MergeToRight
                            | '$' -> ColumnOptions.NoFill
                            | 'x' -> ColumnOptions.Truncate
                            | 'X' -> ColumnOptions.TruncateV2
                            | '#' -> ColumnOptions.NoColSep
                            | _ -> ColumnOptions.Default

                        if flag <> ColumnOptions.Default then
                            result ||| flag
                        else
                            result)
                    ColumnOptions.Default

        let width = int matchResult.Groups[2].Value

        let ansi =
            if not matchResult.Groups[5].Success then
                String.Empty
            else
                matchResult.Groups[5].Value

        { Width = width
          Justification = justification
          Options = options
          Ansi = ansi }

    /// <summary>
    /// Parses a space-separated list of column specifications.
    /// </summary>
    /// <param name="spec">The space-separated column specification string.</param>
    let parseList (spec: string) : ColumnSpec list = spec.Split(' ') |> Seq.map parse |> Seq.toList

File: SharpMUSH.MarkupString/ColumnModule.cs
// Converted from ColumnModule.fs — module SharpMUSH.MarkupString.ColumnModule
using System;
using System.Text.RegularExpressions;

namespace SharpMUSH.MarkupString.ColumnModule;

public enum Justification { Left, Center, Right, Full, Paragraph }

[Flags]
public enum ColumnOptions
{
    Default    = 0,
    Repeat     = 1,
    MergeToLeft  = 2,
    MergeToRight = 4,
    NoFill     = 8,
    Truncate   = 16,
    TruncateV2 = 32,
    NoColSep   = 64,
}

public record ColumnSpec(int Width, Justification Justification, ColumnOptions Options, string Ansi);

public static partial class ColumnSpecParser
{
    [GeneratedRegex(@"^([<>=_\-])?(\d+)([\.`'$xX#]*)(?:\((.+)\))?$")]
    private static partial Regex WidthPattern();

    public static ColumnSpec Parse(string spec)
    {
        var m = WidthPattern().Match(spec);
        if (!m.Success)
            throw new ArgumentException($"Invalid column specification: {spec}");

        Justification justification = m.Groups[1].Success
            ? m.Groups[1].Value switch
            {
                "<" => Justification.Left,
                "=" => Justification.Paragraph,
                ">" => Justification.Right,
                "_" => Justification.Full,
                "-" => Justification.Center,
                _   => Justification.Left,
            }
            : Justification.Left;

        ColumnOptions options = ColumnOptions.Default;
        if (m.Groups[3].Success)
            foreach (char c in m.Groups[3].Value)
                options |= c switch
                {
                    '.'  => ColumnOptions.Repeat,
                    '`'  => ColumnOptions.MergeToLeft,
                    '\'' => ColumnOptions.MergeToRight,
                    '$'  => ColumnOptions.NoFill,
                    'x'  => ColumnOptions.Truncate,
                    'X'  => ColumnOptions.TruncateV2,
                    '#'  => ColumnOptions.NoColSep,
                    _    => ColumnOptions.Default,
                };

        int width = int.Parse(m.Groups[2].Value);
        string ansi = m.Groups[4].Success ? m.Groups[4].Value : string.Empty;

        return new ColumnSpec(width, justification, options, ansi);
    }

    public static System.Collections.Generic.List<ColumnSpec> ParseList(string spec)
    {
        var result = new System.Collections.Generic.List<ColumnSpec>();
        foreach (var token in spec.Split(' '))
            result.Add(Parse(token));
        return result;
    }
}
